cluster-autoscaler/core/scaledown/actuation/update_latency_tracker.go

```diff
@@ -0,0 +1,79 @@
+package actuation
+
+import (
+	"time"
+
+	"k8s.io/autoscaler/cluster-autoscaler/utils/kubernetes"
+	"k8s.io/autoscaler/cluster-autoscaler/utils/taints"
+	"k8s.io/klog/v2"
+)
+
+type UpdateLatencyTracker struct {
+	startTimestamp  map[string]time.Time
+	finishTimestamp map[string]time.Time
+	nodeLister      kubernetes.NodeLister
+}
+
+func NewUpdateLatencyTracker(startTimestamp map[string]time.Time, nodeLister kubernetes.NodeLister) *UpdateLatencyTracker {
+	return &UpdateLatencyTracker{
+		startTimestamp:  startTimestamp,
+		finishTimestamp: map[string]time.Time{},
+		nodeLister:      nodeLister,
+	}
+}
+
+func (u *UpdateLatencyTracker) Start(stopChan <-chan struct{}, resultChan chan int64) {
+	for {
+		select {
+		case <-stopChan:
+			var remainingNodes []string
+			for node, _ := range u.startTimestamp {
+				remainingNodes = append(remainingNodes, node)
+			}
+			for remainingNodes != nil {
+				time.Sleep(100 * time.Millisecond)
+				u.updateFinishTime()
+				remainingNodes = u.remainingNodesToFinish(remainingNodes)
+			}
+			latency := u.calculateLatency()
+			resultChan <- latency
+			return
+		default:
+		}
+		u.updateFinishTime()
+		time.Sleep(50 * time.Millisecond)
+	}
+}
+
+func (u *UpdateLatencyTracker) updateFinishTime() {
+	nodes, err := u.nodeLister.List()
+	if err != nil {
+		klog.Errorf("Error listing nodes: %v", err)
+	}
+	for _, node := range nodes {
+		if _, ok := u.finishTimestamp[node.Name]; !ok && taints.HasToBeDeletedTaint(node) {
+			u.finishTimestamp[node.Name] = time.Now()
+		}
+	}
+}
+
+func (u *UpdateLatencyTracker) remainingNodesToFinish(nodes []string) []string {
+	var remainingNodes []string
+	for _, node := range nodes {
+		if _, ok := u.finishTimestamp[node]; !ok {
+			remainingNodes = append(remainingNodes, node)
+		}
+	}
+	return remainingNodes
+}
+
+func (u *UpdateLatencyTracker) calculateLatency() int64 {
```

Comment:
I'd take max or some high percentile of durations here to be on the safe side. I'd also expect this to return a duration.

Reply:
Done